walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,98 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
-}
+  private static final Collection<Event> NO_EVENTS = Collections.emptySet();
+  private static final Collection<String> NO_ATTENDEES = Collections.emptySet();  
+  private static final int endOfDay = TimeRange.END_OF_DAY; 
+  
+  private static int WHOLE_DAY = 1440;
+
+    // sort and return busy times for requested attendees
+    private static Collection<TimeRange> getCoveredTimesForAttendees(Collection<Event> events,
+    Collection<String> mandatoryAttendees) {   
+      Set<String> setOfAttendees = new HashSet<String>();
+      for(String attendee : mandatoryAttendees) {
+        setOfAttendees.add(attendee);
+      }
+
+      // define 'coveredTimesForAttendees' as ArrayList to use 
+      // suitable method for sorting the time ranges
+      ArrayList<TimeRange> coveredTimesForAttendees = new ArrayList <TimeRange> ();
+      for(Event event: events) {
+        Collection<String> attendeesInEvent = event.getAttendees();
+        TimeRange timeCoveredByEvent = event.getWhen();
+        for(String attendee : attendeesInEvent) {   
+          if(setOfAttendees.contains(attendee)) {
+            coveredTimesForAttendees.add(timeCoveredByEvent);
+            break;
+          } 
+        }
+      }
+      Collections.sort(coveredTimesForAttendees, TimeRange.ORDER_BY_START);
+      return coveredTimesForAttendees;
+    }
+
+    // calculate and return available time ranges for requested events
+    private static Collection<TimeRange> getAvailableTimes(Collection<Event> events,
+    Collection<String> mandatoryAttendees, long requestDuration) {
+      Collection<TimeRange> availableTimes = new ArrayList<TimeRange>();
+
+      int timeForNewEventsBegin = TimeRange.START_OF_DAY;
+      int eventStart = 0;
+      int eventEnd = 0;
+
+      if (requestDuration > WHOLE_DAY) {
+        return availableTimes;
+      }
+
+      if(mandatoryAttendees == NO_ATTENDEES) {
+        availableTimes.add(TimeRange.WHOLE_DAY);
+        return availableTimes;
+      }
+
+      if(events == NO_EVENTS) {
```

Comment:
Hmm I am actually not very sure if this works. If all your purpose is to check a collection is empty, it would be much better to do events.isEmpty(). That is much more straightforward. Plus you are comparing the reference here, not the actual content.

Say if I do
  ArrayList<Event> another_events = new ArrayList<Event>();
  if (another_events == NO_EVENTS) { / / This will be false because the references are not the same.
  }

The same applies to NO_ATTENDEES.

Reply:
You're right! Done

Reply:
You're right, done!